src/cli/commands/upgrade.js

```diff
@@ -21,53 +67,110 @@ export function hasWrapper(commander: Object, args: Array<string>): boolean {
+  };
+
+  // Attempt to preserve the range operator from the package.json specified semver range.
+  // If an explicit operator was specified using --exact, --tilde, --caret, then that will take precedence.
+  const buildPatternToUpgradeTo = (dep, flags) => {
+    if (dep.latest === 'exotic') {
+      return dep.url;
     }
 
-    if (!flags.scope.endsWith('/')) {
-      flags.scope += '/';
+    const toLatest = flags.latest;
+    const toVersion = toLatest ? dep.latest : dep.range;
+    let rangeOperator = '';
+
+    if (toLatest) {
+      if (flags.caret) {
+        rangeOperator = '^';
+      } else if (flags.tilde) {
+        rangeOperator = '~';
+      } else if (flags.exact) {
+        rangeOperator = '';
+      } else {
+        rangeOperator = getRangeOperator(dep.range);
+      }
     }
 
-    if (/^@[a-zA-Z0-9-][a-zA-Z0-9_.-]*\/$/g.test(flags.scope)) {
-      addArgs = Object.keys(allDependencies)
-        .filter(dependency => {
-          return dependency.startsWith(flags.scope);
-        })
-        .map(dependency => {
-          return getDependency(allDependencies, dependency);
-        });
-    } else {
-      throw new MessageError(reporter.lang('scopeNotValid'));
+    return `${dep.name}@${rangeOperator}${toVersion}`;
+  };
+
+  // ensure scope is of the form `@scope/`
+  const normalizeScope = function() {
+    if (flags.scope) {
+      if (!flags.scope.startsWith('@')) {
+        flags.scope = '@' + flags.scope;
+      }
+
+      if (!flags.scope.endsWith('/')) {
+        flags.scope += '/';
+      }
     }
-  } else if (flags.latest && args.length === 0) {
-    addArgs = Object.keys(allDependencies).map(dependency => getDependency(allDependencies, dependency));
-  } else {
-    addArgs = args.map(dependency => {
-      return getDependency(allDependencies, dependency);
-    });
-  }
+  };
 
-  const addFlags = Object.assign({}, flags, {force: true});
+  const versionFilter = function(dep: Dependency): boolean {
+    return dep.current !== dep[outdatedFieldName];
+  };
 
-  const install = new Add(addArgs, addFlags, config, reporter, lockfile);
-  await install.init();
-}
+  const scopeFilter = function(dep: Dependency): boolean {
+    if (validScopeRegex.test(flags.scope)) {
+      return dep.name.startsWith(flags.scope);
+    }
 
-function getDependency(allDependencies, dependency): string {
-  const remoteSource = allDependencies[dependency];
+    return true;
+  };
 
-  if (remoteSource && getExoticResolver(remoteSource)) {
-    return remoteSource;
+  if (!flags.latest) {
+    // these flags only have an affect when --latest is used
+    flags.tilde = false;
+    flags.exact = false;
+    flags.caret = false;
   }
 
-  return dependency;
+  normalizeScope();
+
+  const deps = (await PackageRequest.getOutdatedPackages(lockfile, install, config, reporter, patterns))
+    .filter(versionFilter)
+    .filter(scopeFilter);
+  deps.forEach(dep => (dep.upgradeTo = buildPatternToUpgradeTo(dep, flags)));
```

Comment:
It won't block the merge but I think we should try using `for ... of` when possible, instead of `forEach`